server/docker_resource.py

```diff
@@ -0,0 +1,433 @@
+        self.resourceName = name
+        DockerResource.resourceName = name
+        self.route('PUT', (DockerResource.resourceName, 'docker_image'),
+                   self.appendImage)
+        self.route('DELETE', (DockerResource.resourceName, 'docker_image'),
+                   self.deleteImage)
+        self.route('GET', (DockerResource.resourceName, 'docker_image'),
+                   self.getDockerImages)
+
+    @access.admin
+    @describeRoute(
+        Description('list docker images and their clis ')
+        .notes("""Must be a system administrator to call this. If the value
+                       passed is a valid JSON object.""")
+        .errorResponse(
+            'You are not a system administrator.', 403)
+        .errorResponse(
+            'Failed to set system setting.', 500)
+    )
+    def getDockerImages(self, params):
+        data = {}
+        current = DockerCache(getDockerImageSettings())
+        for image in current.getDockerImageList():
+            data[image] = current.getCLIDict(image)
+        return data
+
+    @access.admin
+    @describeRoute(
+        Description('Remove a docker image ')
+        .notes(
+            """Must be a system administrator to call this. If the value
+                       passed is a valid JSON object.""")
+        .param('name', 'The name or a list of names  of the '
+                       'docker images to be removed', required=True)
+        .errorResponse('You are not a system administrator.', 403)
+        .errorResponse('Failed to set system setting.', 500)
+    )
+    # TODO delete REST endpoint
+    def deleteImage(self, params):
+
+        self.requireParams(('name',), params)
+        name = params['name']
+        name = json.loads(name)
+
+        if isinstance(name, list):
+            for img in name:
+
+                if not isinstance(img, six.string_types):
+                    raise RestException('%s was not a valid string.' % img)
+            else:
+                self._deleteImage(name)
+        elif isinstance(name, six.string_types):
+
+            self._deleteImage([name])
+
+        else:
+
+            raise RestException('name was not a valid JSON list or string.')
+
+    def _deleteImage(self, names):
+        """
+        Removes the docker images and there respective clis from the settings
+        :param name: The name of the docker image (user/rep:tag)
+
+        """
+        currentSettings = DockerCache(getDockerImageSettings())
+        for name in names:
+
+            if not currentSettings.deleteImage(name):
+                raise RestException('%s does not exist' % name)
+        # need to remove each cli attribute
+        # need to remove the clli list attribute
+        # need to remove the rest route
+        print "after delete", currentSettings.raw()
+        ModelImporter.model('setting').set(PluginSettings.DOCKER_IMAGES,
+                                           currentSettings.raw())
```

Comment:
Each docker image may have multiple cli's. The server exposes each cli as a rest endpoint. Originally the server would query for all cli's available and their xml specs upon loading the histomicstk plugin. This requires multiple executions of docker run and delayed the loading of the  histomicstk plugin. The actual cli was executed as an async job when the rest endpoint is being used.